protocol/ibd/ibd.go

```diff
@@ -0,0 +1,244 @@
+package ibd
+
+import (
+	"github.com/kaspanet/kaspad/blockdag"
+	"github.com/kaspanet/kaspad/netadapter/router"
+	"github.com/kaspanet/kaspad/protocol/common"
+	peerpkg "github.com/kaspanet/kaspad/protocol/peer"
+	"github.com/kaspanet/kaspad/protocol/protocolerrors"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"sync"
+	"sync/atomic"
+)
+
+var (
+	isIBDRunning  uint32
+	startIBDMutex sync.Mutex
+)
+
+// StartIBDIfRequired selects a peer and starts IBD against it
+// if required
+func StartIBDIfRequired(dag *blockdag.BlockDAG) error {
+	startIBDMutex.Lock()
+	defer startIBDMutex.Unlock()
+
+	if atomic.LoadUint32(&isIBDRunning) != 0 {
+		return nil
+	}
+
+	peer, err := selectPeerForIBD(dag)
+	if err != nil {
+		return err
+	}
+	if peer == nil {
+		return requestSelectedTipsIfRequired(dag)
+	}
+
+	atomic.StoreUint32(&isIBDRunning, 1)
+	peer.StartIBD()
+	return nil
+}
+
+// selectPeerForIBD returns the first peer whose selected tip
+// hash is not in our DAG
+func selectPeerForIBD(dag *blockdag.BlockDAG) (*peerpkg.Peer, error) {
+	for _, peer := range peerpkg.ReadyPeers() {
+		peerSelectedTipHash, err := peer.SelectedTipHash()
+		if err != nil {
+			return nil, err
+		}
+		if !dag.IsInDAG(peerSelectedTipHash) {
+			return peer, nil
+		}
+	}
+	return nil, nil
+}
+
+// HandleIBD waits for IBD start and handles it when IBD is triggered for this peer
+func HandleIBD(incomingRoute *router.Route, outgoingRoute *router.Route,
+	peer *peerpkg.Peer, dag *blockdag.BlockDAG) error {
+
+	for {
+		peer.WaitForIBDStart()
+
+		// We run the flow inside a func so that the defer is called at its end
```

Comment:
I suggest creating a separate func, so that this one is only flow control (waiting for start, then calling the actual flow)